env: move HOME away from buildpath/testpath

DomT4 opened the pull request:
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

Think I've been talking about this with @ilovezfs for _years_, so here's an actual PR. Likely obvious but even if you really really love it as-is it'll need matching `homebrew/core` PRs before this can be rolled in.

Keeping `HOME` inside the buildpath/testpath has led to a series of required silliness in core such as:
* https://github.com/Homebrew/homebrew-core/blob/2e34a6cdb183c0a43f79b777c0a46eca3111ce5e/Formula/docker-credential-helper.rb#L21
* https://github.com/Homebrew/homebrew-core/blob/5b3e828f26db0b73a534d401c71f0f0fd54db2c3/Formula/git-annex.rb#L72-L73
* https://github.com/Homebrew/homebrew-core/blob/4f66111d393034cbed4c50d5fed47e51422d179e/Formula/openshift-cli.rb#L23-L24

For the sake of a fun little piece of history, ref: https://github.com/Homebrew/brew/commit/ac523bac06.

ilovezfs commented:
`#{buildpath}/../brew_home` might be a little nicer than yet another non-deterministic temp directory

ilovezfs commented:
If we don't end up doing something like `#{buildpath}/../brew_home`, I think HOME should at least have the formula name (and probably version) in it so we don't have a collection of anonymous `/tmp/brew_home-*`.

ilovezfs commented:
Also, if we don't do something deterministic like `#{buildpath}/../brew_home`, then we should set an envvar with the buildpath so that if you `cd $HOME` while in debrew, it's not difficult to find your way back to the buildpath.

MikeMcQuaid commented on Library/Homebrew/formula.rb:
```diff
@@ -1690,8 +1690,9 @@ def run_test
     mktemp("#{name}-test") do |staging|
       staging.retain! if ARGV.keep_tmp?
       @testpath = staging.tmpdir
+      @env_home = Pathname.new(Dir.mktmpdir("brew_home-", HOMEBREW_TEMP))
```
The `_` and `-` in the same name is making me 😭 :trollface:

DomT4 replied:
See, now that you've mentioned it, it's going to bother me too 😬.

MikeMcQuaid commented:
> #{buildpath}/../brew_home might be a little nicer than yet another non-deterministic temp directory

Agreed with something like this (and the general idea of the change) 👍

DomT4 commented:
> #{buildpath}/../brew_home might be a little nicer than yet another non-deterministic temp directory

I'd prefer not to go down this path because of it being possible to build multiple things at once in `brew`; in the name of sandboxing & sanity I don't want things to end up sharing a `HOME` directory.

> I think HOME should at least have the formula name (and probably version) in it so we don't have a collection of anonymous 

I'm happier to roll with this sort of thing though given that'll still be unique, if you're happier with that option over the existing one.

DomT4 commented:
> We're already in a unique temporary directory like

Heh, this is what I get for replying before the morning coffee. Forgot for a second we use `name-version` under the initial temporary directory 🤦‍♂️. Your initial point was correct/valid, my apologies. I'll make some changes.

ilovezfs commented:
No worries.

Personally I think just making sure there is at least one subdir is nicer than always creating `build`.

(Note besides the nounzip case, there are also a few rare (and "wrong") tarballs for the regular strategy that extract their contents in the current directory instead of a subdir.)

DomT4 commented:
Alright. _Not sure_ how much you're going to love what I've done here but there's some changes to review 🙈.

MikeMcQuaid commented on Library/Homebrew/download_strategy.rb:
```diff
@@ -447,7 +447,12 @@ def _fetch
+    # brew_home ends up in HOMEBREW_TEMP rather than the (build|test)path.
+    mkdir "staging"
+    cp cached_location, "staging/#{basename_without_params}", preserve: true
+    cd "staging"
```
Do other strategies do the `cd`?

DomT4 replied:
_Inside_ the buildpath/testpath itself? _I think this might_ be unique. I'm essentially doing it this way to preserve backwards compatibility with formulae without requiring changes on that side of things for this aspect, whilst still allowing us to force everything once subdirectory deep.

Bit of a dance maybe but so far what I've tested works as expected.

ilovezfs replied:
not unique. idiomatic way is `chdir`
See https://github.com/Homebrew/brew/blob/master/Library/Homebrew/download_strategy.rb#L250 and https://github.com/Homebrew/brew/blob/master/Library/Homebrew/download_strategy.rb#L269-L279

MikeMcQuaid replied:
Cool, thanks both 👍

DomT4 replied:
> idiomatic way is chdir

Ah, of course 👍. That automatic jump didn't seem to work for `nounzip`, but that seems to be expected; we'd presumably need a `when 2 then` but don't know if you're happy with where the logic is now or whether you want a more general case.

ilovezfs replied:
maybe a function for handling nounzip and tarball-extracting-to-current-dir and whatever else comes up?

ilovezfs replied:
if child count is 0, empty archive
elsif child count is 1 and child is a dir, cd into dir
elsif one of the children is named staging already, set the house on fire
else
mkdir staging and move children - staging into staging and cd into staging

ilovezfs replied:
(could probably live in chdir or be called from there)

DomT4 replied:
Yeah, that sounds like a viable option, I'll look into it.

ilovezfs commented:
I think we need to be able to access the home as a Pathname during the build and during the test block so that we're not running around doing `Pathname.new("#{ENV["HOME"]}")` everywhere.

ilovezfs commented:
i.e. it needs `attr_reader :env_home`

Also wondering if `home` instead of `env_home` might be a nicer name.

ilovezfs commented:
In addition, I'd _really_ like to see something here that raises an exception if HOME is ever equal to `HOMEBREW_TEMP/brew_home` as I think nounzip may not be the only strategy affected. For example, does this code correctly handle the case where a tarball extracts to the current directory?

DomT4 commented:
> i.e. it needs attr_reader :env_home

Sure, happy to look at doing it this way.

> Also wondering if home instead of env_home might be a nicer name.

I think at the time I wrote it as `env_home` to make it clearer at a glance we're not "violating" `HOME` itself if we ever need to do things like `home.write <<~EOS` or `home.install`, but I'm not wedded to that logic if you're not fussed.

> In addition, I'd really like to see something here that raises an exception if HOME is ever equal to HOMEBREW_TEMP/brew_home as I think nounzip may not be the only strategy affected.

Sure, that seems like a good step in the name of safety.

> does this code correctly handle the case where a tarball extracts to the current directory?

So far my testing formulae have been `apktool`, `a2ps`, `neovim` and `vis`, both build & test. I'll try and broaden the pool a bit as we move forwards here but shout if there's anything specific you want it tested against.

MikeMcQuaid commented:
> I think at the time I wrote it as env_home to make it clearer at a glance we're not "violating" HOME itself if we ever need to do things like home.write <<~EOS or home.install, but I'm not wedded to that logic if you're not fussed.

`env_home` has grown on me a bit.

ilovezfs commented:
>I think at the time I wrote it as env_home to make it clearer at a glance we're not "violating" HOME itself if we ever need to do things like home.write <<~EOS or home.install, but I'm not wedded to that logic if you're not fussed.

What does "HOME itself" mean here?

DomT4 commented:
> What does "HOME itself" mean here?

`/Users/Joe`, for example.

ilovezfs commented:
sandbox much?

ilovezfs commented:
yup as expected
```
iMac-TMP:homebrew-core joe$ brew install -dvs hapi-fhir-cli
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/hapi-fhir-cli.rb
/usr/bin/sandbox-exec -f /tmp/homebrew20180620-59992-g4i1le.sb nice /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.3.3_2/bin/ruby -W0 -I /usr/local/Homebrew/Library/Homebrew:/usr/local/Homebrew/Library/Homebrew/cask/lib -- /usr/local/Homebrew/Library/Homebrew/build.rb /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/hapi-fhir-cli.rb --verbose --debug
/usr/local/Homebrew/Library/Homebrew/build.rb (Formulary::FromPathLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/hapi-fhir-cli.rb
==> Downloading https://github.com/jamesagnew/hapi-fhir/releases/download/v3.4.0/hapi-fhir-3.4.0-cli.tar.bz2
Already downloaded: /Users/joe/Library/Caches/Homebrew/hapi-fhir-cli-3.4.0.tar.bz2
==> Verifying hapi-fhir-cli-3.4.0.tar.bz2 checksum
tar xjf /Users/joe/Library/Caches/Homebrew/hapi-fhir-cli-3.4.0.tar.bz2
/usr/local/Homebrew/Library/Homebrew/debrew.rb:11:in `raise'
RuntimeError: here
1. raise
2. ignore
3. backtrace
4. irb
5. shell
Choose an action: 5
When you exit this shell, you will return to the menu.
bash-3.2$ echo $HOME
/private/tmp/brew_home
bash-3.2$
```
So tarballs that extract to current directory have same problem as nounzip.

DomT4 commented:
The sandbox is a _relatively_ recent thing in terms of Homebrew's history, especially on by default. Our former CI machines were an excellent testament to how long we allowed anything to write whatever it wanted to `~/`. I'm not sure I'd assume everyone who looks at Homebrew's code will go _"Oh yeah, the sandbox!"_.

DomT4 commented:
> brew install -dvs hapi-fhir-cli

Thanks, I'll check this one out.

ilovezfs commented:
same issue with at least some `--HEAD` builds:
```
iMac-TMP:Homebrew joe$ brew install --HEAD -dv wrangler
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/wrangler.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/erlang.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/openssl.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/autoconf.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/automake.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libtool.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/makedepend.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/pkg-config.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/wxmac.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/jpeg.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libpng.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libtiff.rb
/usr/bin/sandbox-exec -f /tmp/homebrew20180620-62542-168nab.sb nice /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.3.3_2/bin/ruby -W0 -I /usr/local/Homebrew/Library/Homebrew:/usr/local/Homebrew/Library/Homebrew/cask/lib -- /usr/local/Homebrew/Library/Homebrew/build.rb /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/wrangler.rb --verbose --debug --HEAD
/usr/local/Homebrew/Library/Homebrew/build.rb (Formulary::FromPathLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/wrangler.rb
/usr/local/Homebrew/Library/Homebrew/build.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/erlang.rb
/usr/local/Homebrew/Library/Homebrew/build.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/openssl.rb
/usr/local/Homebrew/Library/Homebrew/build.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/wxmac.rb
/usr/local/Homebrew/Library/Homebrew/build.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/jpeg.rb
/usr/local/Homebrew/Library/Homebrew/build.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libpng.rb
/usr/local/Homebrew/Library/Homebrew/build.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libtiff.rb
/usr/local/Homebrew/Library/Homebrew/build.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/autoconf.rb
/usr/local/Homebrew/Library/Homebrew/build.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/automake.rb
/usr/local/Homebrew/Library/Homebrew/build.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libtool.rb
/usr/local/Homebrew/Library/Homebrew/build.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/makedepend.rb
/usr/local/Homebrew/Library/Homebrew/build.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/pkg-config.rb
==> Cloning https://github.com/RefactoringTools/wrangler.git
Updating /Users/joe/Library/Caches/Homebrew/wrangler--git
git config remote.origin.url https://github.com/RefactoringTools/wrangler.git
git config remote.origin.fetch +refs/heads/master:refs/remotes/origin/master
git fetch origin
==> Checking out branch master
git checkout -f master --
Already on 'master'
Your branch is up-to-date with 'origin/master'.
git reset --hard origin/master
HEAD is now at aa2330a Redo behaviour extraction example with size 7
/usr/local/Homebrew/Library/Homebrew/debrew.rb:11:in `raise'
RuntimeError: here
1. raise
2. ignore
3. backtrace
4. irb
5. shell
Choose an action: 5
When you exit this shell, you will return to the menu.
bash-3.2$ echo $HOME
/private/tmp/brew_home
bash-3.2$
```

ilovezfs commented:
(head clones into the top level temp dir. we could probably use repo name instead)

ldemailly commented:
oh awesome, that moves .home_brew out of the source tree right ? that would solve https://github.com/istio/fortio/issues/259

ilovezfs commented:
yes

DomT4 commented:
Next push here is _likely_ to land somewhere between 3-5 hours from now. Probably going to take a couple hours to wrap up a personal thing & then an hour or two after that to rewrite/test this.

ilovezfs commented:
no rush. ⏳

DomT4 commented:
Will be sometime Thursday afternoon now; personal thing is proving more of a time hog than expected. That'll teach me for providing a loose time estimate of when I'd next move on this, heh.

MikeMcQuaid commented:
As a random related comment: will we put anything in the `buildpath` after this (e.g. java cache, anything else I've not thought about)? Also even more unrelated: maybe it's time to think about e.g. a git checkout having `.git` in the buildpath by default.

ilovezfs commented:
java_cache already isn't in buildpath

.git already is

MikeMcQuaid commented:
> .git already is

I mean without needing to manually symlink it (or is my memory shit again)

ilovezfs commented:
It's a full copy. See https://github.com/Homebrew/brew/blob/41e27928c475e2c8ed8c93161269936bd85f8847/Library/Homebrew/download_strategy.rb#L758-L761

DomT4 commented:
> will we put anything in the buildpath after this (e.g. java cache, anything else I've not thought about)?

I think `.brew_home` being in the `buildpath`/`testpath` and the hassle that has caused both us & external projects is a pretty strong motivator to try and keep anything from the Homebrew side of things out of the `(build|test)path` moving forwards, personally.

ilovezfs commented:
I'm not eager to e.g. rewrite language/haskell.rb unless we can point to specific current problems.

DomT4 commented:
Yeah, I looked that over when I was initially drafting this PR but left it be because unlike `.brew_home` no maintainer, contributor, user or upstream project has had to report an issue or implement a workaround.

I'd concur on leaving it be unless it proves problematic, but ideally moving forwards anything else would stay out of the build/testpath where possible.

DomT4 commented:
This still needs a fair bit of testing but there's some code to review at least.

DomT4 commented:
Yeah I broke my regular testing formula (`a2ps`) with the latest changes so more testing is definitely in order 😆. But progress is progress.

ilovezfs commented on Library/Homebrew/formula.rb:
```diff
@@ -1934,26 +1943,29 @@ def stage
         with_env(stage_env) do
+          if (HOMEBREW_TEMP/"brew_home").exist? &&
+            ENV["HOME"] == (HOMEBREW_TEMP/"brew_home").realpath.to_s
+            raise "HOME is #{ENV["HOME"]}; refusing to proceed!"
```
this should be checked before creating the directory

Triggered with
```
iMac-TMP:homebrew-core joe$ ls /tmp/brew_home
ls: /tmp/brew_home: No such file or directory
iMac-TMP:homebrew-core joe$ brew  pull  https://github.com/Homebrew/homebrew-core/pull/29337
==> Fetching patch 
Patch: https://github.com/Homebrew/homebrew-core/pull/29337.patch
######################################################################## 100.0%
==> Applying patch
Applying: unp64 2.35
Warning: unp64 has a bottle: do you need to update it with --bottle?
==> Patch closes issue #29337
==> Patch changed:
 Formula/unp64.rb | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)
iMac-TMP:homebrew-core joe$ brew instal -dvs unp64
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/unp64.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/p7zip.rb
/usr/bin/sandbox-exec -f /tmp/homebrew20180622-1349-1f917cd.sb nice /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.3.3_2/bin/ruby -W0 -I /usr/local/Homebrew/Library/Homebrew:/usr/local/Homebrew/Library/Homebrew/cask/lib -- /usr/local/Homebrew/Library/Homebrew/build.rb /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/unp64.rb --verbose --debug
/usr/local/Homebrew/Library/Homebrew/build.rb (Formulary::FromPathLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/unp64.rb
/usr/local/Homebrew/Library/Homebrew/build.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/p7zip.rb
==> Downloading http://iancoog.altervista.org/C/unp64_235.7z
Already downloaded: /Users/joe/Library/Caches/Homebrew/unp64-2.35.7z
==> Verifying unp64-2.35.7z checksum
7zr x /Users/joe/Library/Caches/Homebrew/unp64-2.35.7z

7-Zip (a) [64] 16.02 : Copyright (c) 1999-2016 Igor Pavlov : 2016-05-21
p7zip Version 16.02 (locale=utf8,Utf16=on,HugeFiles=on,64 bits,8 CPUs x64)

Scanning the drive for archives:
1 file, 277047 bytes (271 KiB)               

Extracting archive: /Users/joe/Library/Caches/Homebrew/unp64-2.35.7z
--
Path = /Users/joe/Library/Caches/Homebrew/unp64-2.35.7z
Type = 7z
Physical Size = 277047
Headers Size = 4551
Method = LZMA2:16 BCJ
Solid = +
Blocks = 2

Everything is Ok

Folders: 4
Files: 242
Size:       1280491
Compressed: 277047
Error: HOME is /private/tmp/brew_home; refusing to proceed!
/usr/local/Homebrew/Library/Homebrew/formula.rb:1967:in `block (2 levels) in stage'
/usr/local/Homebrew/Library/Homebrew/utils.rb:557:in `with_env'
/usr/local/Homebrew/Library/Homebrew/formula.rb:1964:in `block in stage'
/usr/local/Homebrew/Library/Homebrew/resource.rb:119:in `block in unpack'
/usr/local/Homebrew/Library/Homebrew/extend/fileutils.rb:13:in `block in mktemp'
/usr/local/Homebrew/Library/Homebrew/extend/fileutils.rb:73:in `block in run'
/usr/local/Homebrew/Library/Homebrew/extend/fileutils.rb:73:in `chdir'
/usr/local/Homebrew/Library/Homebrew/extend/fileutils.rb:73:in `run'
/usr/local/Homebrew/Library/Homebrew/extend/fileutils.rb:12:in `mktemp'
/usr/local/Homebrew/Library/Homebrew/resource.rb:114:in `unpack'
/usr/local/Homebrew/Library/Homebrew/resource.rb:92:in `stage'
/usr/local/Homebrew/Library/Homebrew/formula.rb:1943:in `stage'
/usr/local/Homebrew/Library/Homebrew/formula.rb:1116:in `brew'
/usr/local/Homebrew/Library/Homebrew/build.rb:113:in `block in install'
/usr/local/Homebrew/Library/Homebrew/utils.rb:557:in `with_env'
/usr/local/Homebrew/Library/Homebrew/build.rb:110:in `install'
/usr/local/Homebrew/Library/Homebrew/build.rb:191:in `<main>'
iMac-TMP:homebrew-core joe$ ls /tmp/brew_home
Library
```

DomT4 replied:
Both points addressed. The `download_strategy` changes in particular still need _a lot_ of rewriting and testing.

DomT4 commented:
Still working out how to handle stuff that clones from `git`/`svn`/etc here, which is the delay. Thus far I haven't yet come up with something that works as well as I would hope, not least with relatively complex formula like [cling](https://github.com/Homebrew/homebrew-core/blob/dce4485a9f32da74b1f1b0a2f77b488a708bb4be/Formula/cling.rb).

Everything else _seems_ to be working but I'm assuming we don't want to go and deprecate anything that comes from a `git clone` or `svn checkout` or such 😉, so still working on it.

ilovezfs commented:
How is cling especially problematic?

DomT4 commented:
> How is cling especially problematic?

Well, blaming `cling` individually may be harsh, but it's a decent test formula in terms of complexity & did a good job taking a hammer to my line of thought.

I was hoping the attached diff would do it but instead it seemed to apply it to resources as well as the main formula, and created `brew_home` in `/tmp` regardless of where I wedged the `chdir`.

Before:
```
.
├── build
│   └── CMakeFiles
└── src
    ├── bindings
    ├── cmake
    ├── docs
    ├── examples
    ├── include
    ├── lib
    ├── projects
    ├── resources
    ├── runtimes
    ├── test
    ├── tools
    ├── unittests
    └── utils

16 directories
```

After:
```
.
├── build
└── src
    ├── cling--llvm
    │   ├── bindings
    │   ├── cmake
    │   ├── docs
    │   ├── examples
    │   ├── include
    │   ├── lib
    │   ├── projects
    │   ├── resources
    │   ├── runtimes
    │   ├── test
    │   ├── tools
    │   ├── unittests
    │   └── utils
    └── tools
        ├── clang
        └── cling

19 directories
```

It's been a while since I've touched Brew/Ruby properly with uni/life/etc and doing so again is making me feel quite the idiot 😅.

```diff
diff --git a/Library/Homebrew/download_strategy.rb b/Library/Homebrew/download_strategy.rb
index fb227c80e..8370cf17e 100644
--- a/Library/Homebrew/download_strategy.rb
+++ b/Library/Homebrew/download_strategy.rb
@@ -214,6 +214,8 @@ class VCSDownloadStrategy < AbstractDownloadStrategy
 
   def update; end
 
+  def chdir; end
+
   def current_revision; end
 
   def extract_ref(specs)
@@ -776,7 +779,8 @@ class GitDownloadStrategy < VCSDownloadStrategy
 
   def stage
     super
-    cp_r File.join(cached_location, "."), Dir.pwd, preserve: true
+    cp_r File.join(cached_location, "."), Pathname.pwd/name, preserve: true
+    chdir
   end
 
   def source_modified_time
```

For now I've smacked pause whilst https://github.com/Homebrew/brew/pull/4402 is pushed through because Markus is working quicker than I am & one of us is going to cause a merge conflict for the other, so it may as well be them doing it to me.

DomT4 commented:
Yes bot, I'm aware. Just haven't wanted to play recurring leapfrog with the other `download`/`unpack` strategy work.

DomT4 commented:
(No tangible changes in that push. Just dropped the `download_strategy` changes and rebased so it's easier to pick this back up fresh sooner rather than later).

MikeMcQuaid commented:
@DomT4 What's the latest here?

DomT4 commented:
I was waiting for Markus' work around download/unpacking to be completed so we wouldn't be stepping on each other constantly, but I also kinda got mentally worn out by how much this drifted from my originally very simple change into something where I have to do a battery of testing locally against every time I tweak something so I've backed off it a bit since, in the hope that sooner rather than later I rediscover the desire to get it finished rather than looking at it and going "Oh no, not that PR" :laughing:.

MikeMcQuaid commented:
@DomT4 Cool, no worries 👍
